diff --git a/sdk/ble_stack/profiles/wss/wsss/src/wsss_task.c b/sdk/ble_stack/profiles/wss/wsss/src/wsss_task.c
index ddf22f2..cf92578 100644
--- a/sdk/ble_stack/profiles/wss/wsss/src/wsss_task.c
+++ b/sdk/ble_stack/profiles/wss/wsss/src/wsss_task.c
@@ -120,8 +120,11 @@ static int wsss_meas_send_req_handler(ke_msg_id_t const msgid,
     ASSERT_ERROR(wsss_env);
     uint8_t conidx = KE_IDX_GET(src_id);
 
-
-    if (ke_state_get(dest_id) == WSSS_IDLE)
+    if (gapc_get_conhdl(conidx) == GAP_INVALID_CONHDL)
+    {
+        status = PRF_ERR_DISCONNECTED;
+    }
+    else if (ke_state_get(dest_id) == WSSS_IDLE)
     {
         if (WSSS_IS_ENABLED(WSSS_WT_MEAS_IND_CFG))
         {
@@ -153,7 +156,16 @@ static int wsss_meas_send_req_handler(ke_msg_id_t const msgid,
         wsss_meas_send_cfm_send(conidx, status);
         msg_status = KE_MSG_CONSUMED;
     }
+    else
+    {
+        msg_status = KE_MSG_SAVED;
+    }
 
+    if (status != GAP_ERR_NO_ERROR)
+    {
+        // Send error message to application
+        wsss_meas_send_cfm_send(conidx, status);
+    }
 
     return msg_status;
 }
